File: src/main/java/net/internetisalie/lunar/luacats/lang/psi/impl/LuaCatsLazyCommentImpl.java
package net.internetisalie.lunar.luacats.lang.psi.impl;


import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.impl.source.tree.LazyParseablePsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import net.internetisalie.lunar.luacats.lang.lexer.LuaLazyElementTypes;
import net.internetisalie.lunar.luacats.lang.psi.*;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * @author ilyas
 */
public class LuaCatsLazyCommentImpl extends LazyParseablePsiElement implements LuaCatsComment {
    public LuaCatsLazyCommentImpl(CharSequence text) {
        super(LuaLazyElementTypes.LUACATS_COMMENT, text);
    }

    public String toString() {
        return "LuaCatsLazyCommentImpl(" + getNode().getElementType()  + ")"; // + StringUtil.notNullize(owner != null ? owner.toString() : null, "no owner");
    }

    public void accept(@NotNull LuaCatsVisitor visitor) {
        visitor.visitComment(this);
    }

    @Override
    public void accept(@NotNull PsiElementVisitor visitor) {
        if (visitor instanceof LuaCatsVisitor) accept((LuaCatsVisitor)visitor);
        else super.accept(visitor);
    }

    @Override
    @NotNull
    public List<LuaCatsAliasTag> getAliasTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsAliasTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsAsyncTag> getAsyncTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsAsyncTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsCastTag> getCastTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsCastTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsClassTag> getClassTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsClassTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsDeprecatedTag> getDeprecatedTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsDeprecatedTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsDescription> getDescriptionList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsDescription.class);
    }

    @Override
    @NotNull
    public List<LuaCatsDiagnosticTag> getDiagnosticTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsDiagnosticTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsEnumTag> getEnumTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsEnumTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsFieldTag> getFieldTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsFieldTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsGenericTag> getGenericTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsGenericTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsMetaTag> getMetaTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsMetaTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsModuleTag> getModuleTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsModuleTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsNodiscardTag> getNodiscardTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsNodiscardTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsOperatorTag> getOperatorTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsOperatorTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsOverloadTag> getOverloadTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsOverloadTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsPackageTag> getPackageTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsPackageTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsParamTag> getParamTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsParamTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsPrivateTag> getPrivateTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsPrivateTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsProtectedTag> getProtectedTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsProtectedTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsReturnTag> getReturnTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsReturnTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsSeeTag> getSeeTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsSeeTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsSourceTag> getSourceTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsSourceTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsTypeOption> getTypeOptionList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsTypeOption.class);
    }

    @Override
    @NotNull
    public List<LuaCatsTypeTag> getTypeTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsTypeTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsVarargTag> getVarargTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsVarargTag.class);
    }

    @Override
    @NotNull
    public List<LuaCatsVersionTag> getVersionTagList() {
        return PsiTreeUtil.getChildrenOfTypeAsList(this, LuaCatsVersionTag.class);
    }

}

File: src/main/kotlin/net/internetisalie/lunar/luacats/lang/psi/impl/LuaCatsLazyCommentImpl.kt
package net.internetisalie.lunar.luacats.lang.psi.impl

import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.impl.source.tree.LazyParseablePsiElement
import com.intellij.psi.util.PsiTreeUtil
import net.internetisalie.lunar.lang.psi.LuaLazyElementTypes
import net.internetisalie.lunar.luacats.lang.psi.*

/**
 * @author ilyas
 *
 * The lazy-parseable chameleon node whose single content child is the inner `comment`
 * ([LuaCatsComment], generated as [LuaCatsCommentImpl]) produced by `parseContents`
 * ([net.internetisalie.lunar.lang.psi.LuaLazyElementTypes.LUACATS_COMMENT]). Because the tags and
 * top-level descriptions are direct children of that inner comment (bnf `private commentLine`),
 * every accessor delegates to it. This restores the generated direct-children contract: a
 * description nested inside a tag (`classTag ::= … description?`) no longer leaks into
 * [getDescriptionList], fixing the duplicate-summary and `isDocCommentEmpty` skew (MAINT-27 #38).
 */
class LuaCatsLazyCommentImpl(text: CharSequence?) : LazyParseablePsiElement(LuaLazyElementTypes.LUACATS_COMMENT, text),
    LuaCatsComment {
    override fun toString(): String {
        return "LuaCatsLazyCommentImpl(" + node.elementType + ")"
    }

    fun accept(visitor: LuaCatsVisitor) {
        visitor.visitComment(this)
    }

    override fun accept(visitor: PsiElementVisitor) {
        if (visitor is LuaCatsVisitor) accept(visitor)
        else super.accept(visitor)
    }

    private fun innerComment(): LuaCatsComment? =
        PsiTreeUtil.getChildOfType(this, LuaCatsComment::class.java)

    private fun <T> delegate(select: (LuaCatsComment) -> List<T>): List<T> =
        innerComment()?.let(select) ?: emptyList()

    override fun getAliasTagList(): List<LuaCatsAliasTag> = delegate { it.aliasTagList }

    override fun getAsyncTagList(): List<LuaCatsAsyncTag> = delegate { it.asyncTagList }

    override fun getCastTagList(): List<LuaCatsCastTag> = delegate { it.castTagList }

    override fun getClassTagList(): List<LuaCatsClassTag> = delegate { it.classTagList }

    override fun getDeprecatedTagList(): List<LuaCatsDeprecatedTag> = delegate { it.deprecatedTagList }

    override fun getDescriptionList(): List<LuaCatsDescription> = delegate { it.descriptionList }

    override fun getDiagnosticTagList(): List<LuaCatsDiagnosticTag> = delegate { it.diagnosticTagList }

    override fun getEnumTagList(): List<LuaCatsEnumTag> = delegate { it.enumTagList }

    override fun getFieldTagList(): List<LuaCatsFieldTag> = delegate { it.fieldTagList }

    override fun getGenericTagList(): List<LuaCatsGenericTag> = delegate { it.genericTagList }

    override fun getMetaTagList(): List<LuaCatsMetaTag> = delegate { it.metaTagList }

    override fun getModuleTagList(): List<LuaCatsModuleTag> = delegate { it.moduleTagList }

    override fun getNodiscardTagList(): List<LuaCatsNodiscardTag> = delegate { it.nodiscardTagList }

    override fun getOperatorTagList(): List<LuaCatsOperatorTag> = delegate { it.operatorTagList }

    override fun getOverloadTagList(): List<LuaCatsOverloadTag> = delegate { it.overloadTagList }

    override fun getPackageTagList(): List<LuaCatsPackageTag> = delegate { it.packageTagList }

    override fun getParamTagList(): List<LuaCatsParamTag> = delegate { it.paramTagList }

    override fun getPrivateTagList(): List<LuaCatsPrivateTag> = delegate { it.privateTagList }

    override fun getProtectedTagList(): List<LuaCatsProtectedTag> = delegate { it.protectedTagList }

    override fun getReturnTagList(): List<LuaCatsReturnTag> = delegate { it.returnTagList }

    override fun getSeeTagList(): List<LuaCatsSeeTag> = delegate { it.seeTagList }

    override fun getSourceTagList(): List<LuaCatsSourceTag> = delegate { it.sourceTagList }

    override fun getTypeOptionList(): List<LuaCatsTypeOption> = delegate { it.typeOptionList }

    override fun getTypeTagList(): List<LuaCatsTypeTag> = delegate { it.typeTagList }

    override fun getVarargTagList(): List<LuaCatsVarargTag> = delegate { it.varargTagList }

    override fun getVersionTagList(): List<LuaCatsVersionTag> = delegate { it.versionTagList }
}
